### PR TITLE
Updated /ball to list previous Balls

### PR DESCRIPTION
**Pull request by TheoXIII**
/ball has been changed to a page which explains briefly what a Ball is. The previous balls (currently only 2022 and 2020) are then listed with links to their pages embedded in their pictures. I have added _data/ball/general.yaml to contain the data for the Balls.

Let me know if the Ball pictures or the shamelessly reused joke are found to be objectionable.

### CODE REVIEW

**Comment by LMBishop** [comment]
Maybe a gallery of photos from each year is more appropriate (rather than just one), I don't think it's clear that each section is actually a link to the respective ball page. Also, the header for each year and location is a bit large.

I think this is a lot better than just having the redirect though

**Comment by MattyTheHacker** [comment]
I agree on all the above - just want to echo the need for the navigation to be as clear as possible. Maybe the ball button on the nav bar could be a drop-down showing each year? or maybe something on the page itself? either way just having the images be the links isn't ideal. 

Also agree on having a gallery for each respective ball but I think the links to those should be contained on their respective pages, not on the ball home page.

File: _data/ball/general.yaml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    alt: 'Image of CSS Ball 2022'
+    attendees: '173'
+  - name: 'CSS BALL'
+    year: '2020'
+    location: 'The Malmaison'
+    date: '12th March'
+    link: '/ball/2020'
+    image: '/assets/images/ball-20-dancing.jpg'
+    alt: 'Image of CSS Ball 2020'
+    attendees: '130'
```
Comment (LMBishop) [inline _data/ball/general.yaml:17]: Duplicating this information might be a bit of an issue, is it not possible to include it with or extend the normal [year].yml files?

**Review by JTWWilson — CHANGES_REQUESTED** [review]
Looks pretty good! I think it’s non-obvious that you have to click on the photos to get to the page about each of the year’s Balls so maybe make the title a link too and perhaps have the formatting underline it to make that clear?

There’s not that much point writing more about past Balls given basically no one cares about what happened years ago but I think it would be nice to include the 2020 virtual ball, even if it was a bit weak. A paragraph about how we hosted it in a different way despite COVID and a pic of one of the tickets would be cool. Similarly, I think the Ball/2021 page should either have a few more photos and a couple more sentences on it or just not exist because it looks quite barren at the moment. We could get photos from the 2018 Ball at the Jamhouse in the Jewellery Quarter too. 

Oo also a Ball TeX thrown in there would be nice :)

**Comment by TheoXIII** [comment]
> I agree on all the above - just want to echo the need for the navigation to be as clear as possible. Maybe the ball button on the nav bar could be a drop-down showing each year? or maybe something on the page itself? either way just having the images be the links isn't ideal.
>Also agree on having a gallery for each respective ball but I think the links to those should be contained on their respective pages, not on the ball home page.

I think that a drop-down is a better idea than a page listing the various balls - I'll work on that. Thanks for suggesting.

**Comment by rainestormee** [comment]
> I liked Matty's idea so I've ditched the /ball page for now and turned it back into a redirect. I've added drop-down support - the new way of adding drop downs and links in _config.yml should be intuitive. The 2022 and 2020 balls are now linked under the "Ball" drop-down and I've added a link to the 2020 Ball gallery to the 2020 Ball page. Please remember to review the mobile version - the drop-down should be working there as well, but slightly differently.
> 
> The only thing I need help with right now is with correcting the misalignment between drop-downs and links in the navbar. I've tried changing a lot of CSS fields but had no luck so far. Could someone let me know how to fix this before approving?

I mean failing not being able to fix this, you could always make the /ball page have links to all of the previous years. Not an ideal solution but it solves the dropdown issue. I'll take a look though when I have some time unless someone gets to it first.

**Review by rainestormee — CHANGES_REQUESTED** [review]
Would also like to see information for 2021 Ball if possible. Might need to dig around for screenshots ect.

**Comment by TheoXIII** [comment]
> I mean failing not being able to fix this, you could always go with the original idea of making the /ball page have links to all of the previous years. Not an ideal solution but it solves the dropdown issue. I'll take a look though when I have some time unless someone gets to it first.

I do like the idea of the list page - I'll work on a few changes. I'm just not sure how needed it is if we have a dropdown since it wouldn't have any extra information anyway. I think we will be able to fix the dropdown since it is only a minor problem and I think it's just an issue of finding the exact setting which needs to be tweaked - it's just I'm not familiar with the website's CSS.

Since I obviously wasn't at Ball 2021, if someone could send me a run-down of what went on, along with some screenshots, that would be great, so I can make a page for that.

**Comment by rainestormee** [comment]
> > I mean failing not being able to fix this, you could always go with the original idea of making the /ball page have links to all of the previous years. Not an ideal solution but it solves the dropdown issue. I'll take a look though when I have some time unless someone gets to it first.
> 
> I do like the idea of the list page - I'll work on a few changes. I'm just not sure how needed it is if we have a dropdown since it wouldn't have any extra information anyway. I think we will be able to fix the dropdown since it is only a minor problem and I think it's just an issue of finding the exact setting which needs to be tweaked - it's just I'm not familiar with the website's CSS.
> 
> Since I obviously wasn't at Ball 2021, if someone could send me a run-down of what went on, along with some screenshots, that would be great, so I can make a page for that.

TLDR ball '21 was online, we handed out JustEat? food vouchers and people could buy their own food. The food voucher was greater than the ticket price. It was hosted on this godawful online thing which looked like 2d pokemon but automatically turned on your camera and mic with people around you, there might be some stuff in the drive or Discord archive channels

**Comment by TheoXIII** [comment]
> > > I mean failing not being able to fix this, you could always go with the original idea of making the /ball page have links to all of the previous years. Not an ideal solution but it solves the dropdown issue. I'll take a look though when I have some time unless someone gets to it first.
> > 
> > 
> > I do like the idea of the list page - I'll work on a few changes. I'm just not sure how needed it is if we have a dropdown since it wouldn't have any extra information anyway. I think we will be able to fix the dropdown since it is only a minor problem and I think it's just an issue of finding the exact setting which needs to be tweaked - it's just I'm not familiar with the website's CSS.
> > Since I obviously wasn't at Ball 2021, if someone could send me a run-down of what went on, along with some screenshots, that would be great, so I can make a page for that.
> 
> TLDR ball '21 was online, we handed out JustEat? food vouchers and people could buy their own food. The food voucher was greater than the ticket price. It was hosted on this godawful online thing which looked like 2d pokemon but automatically turned on your camera and mic with people around you, there might be some stuff in the drive or Discord archive channels

Ok sounds good

File: css/main.scss [hunk 2]
```diff
@@ -160,32 +160,67 @@ header {
...
-        height: 0;
-        display: block;
-        background-color: white;
+  }
 
-        transition: width 0.2s;
-      }
+  .dropdown {
+    overflow: hidden;
+  }
```
Comment (LMBishop) [inline css/main.scss:167]:
```suggestion
```

This fixes your alignment problem, but I don't know what implications it will have for mobile

Comment (TheoXIII) [reply css/main.scss:167]: Ok thanks - I'll give it a go.

**Comment by TheoXIII** [comment]
Alignment issues are fixed thanks to Leonardo. The mobile version works fine still.

**Comment by TheoXIII** [comment]
Feel free to take a look at the Ball 2021 page I've added. Note that some of the images may need changing due to data privacy concerns.

**Comment by LMBishop** [comment]
> Feel free to take a look at the Ball 2021 page I've added.

image

Might be a bit of an issue

**Comment by TheoXIII** [comment]
> > Feel free to take a look at the Ball 2021 page I've added.
> 
> image
> 
> Might be a bit of an issue

Whoops, 2021.yaml was missing

File: css/ball/main.scss [hunk 3]
```diff
@@ -121,6 +121,14 @@ body {
         }
     }
 
+    .dropdown {
+      font-family: $font-secondary;
+    }
+
+    .dropdown-content {
+      background-color: $secondary-lowlight;
+    }
```
Comment (LMBishop) [inline css/ball/main.scss:130]:
```suggestion
    .dropdown-content {
      background-color: $secondary-lowlight;
      
      a {
        font-family: $font-secondary;
      }
    }
```

Just to make the font of the dropdown menu the same as the rest of the ball pages

File: _includes/header.html [hunk 4]
```diff
@@ -6,18 +6,44 @@
...
     {% for item in site.menu %}
+    {% if item.children %}
+    <div class="nav dropdown">
+    <div class="dropdown-btn">
+    <button>{{ item.name | capitalize }}</button>
+    </div>
+    <div class="dropdown-content">
+      {% for sub in item.children %}
+      {% assign link = item.name | append: '/' | append: sub %}
+      <a class="nav" href="{{ link | relative_url }}">{{ sub }}</a>
```
Comment (LMBishop) [inline _includes/header.html:17]: I've noticed the annoying twitching whenever you hover over the menu elements in the dropdown because of the fancy animation thing whenever you hover over stuff in the navbar. 

I've looked into it and it seems pretty annoying to fix, so to me the easiest solution is to just remove the `nav` class, but if you can manage to fix it then even better

**Review by LMBishop — CHANGES_REQUESTED** [review]
The 2021 page looks good to me, just a couple of other things I noticed

**Comment by MattyTheHacker** [comment]
not sure if it's me but the drop-down looks off-centre? 

image

**Comment by TheoXIII** [comment]
The off-centre issue with the dropdown has been fixed. To avoid the jitter, I've removed the animation for the dropdown links for now and replaced it with a transition to TeX Purple when you hover over the link.

**Review by LMBishop — APPROVED** [review]

**Comment by MattyTheHacker** [comment]
the 2020 page is quite empty though I don't actually have any info about it to add

**Comment by LMBishop** [comment]
Someone who was actually at the 2020 ball should write those pages. I think this is enough for this PR, and additional things are starting to get out of scope.

**Comment by JTWWilson** [comment]
Agree with Leonardo, this is really good, let's raise more issues and resolve them rather than making this too big

**Review by JTWWilson — APPROVED** [review]

**Review by rainestormee — APPROVED** [review]

**Comment by TheoXIII** [comment]
I've added the page for Ball 2020 - let me know what you think

**Review by MattyTheHacker — APPROVED** [review]

**Comment by rainestormee** [comment]
Nearly got everyone to approve :p

**Comment by LMBishop** [comment]
there was free alcohol?!?
